### ISSUE TITLE
Queries from multiple spawns crash

### ISSUE THREAD

**Issue by benoist** [crystal-lang/crystal-mysql#34]
When I want to query from different spawns using the same connection pool, it crashes for mysql when queries run at the same time. However this works fine for postgres.

```crystal
require "pg"
require "mysql"
require "db"

DB.open("postgres://localhost/postgres") do |db|

  spawn do
    db.query("select 'foo', pg_sleep(1)") do |rs|
      rs.each do
        foo =  rs.read(String)
        timer =  rs.read(Bytes)
      end
    end
    pp "PG foo"
  end

  spawn do
    db.query("select 'bar'") do |rs|
      rs.each do
        bar =  rs.read(String)
      end
    end
    pp "PG bar"
  end
  sleep 2
end

DB.open("mysql://localhost") do |db|
  spawn do
    db.query("select 'foo', sleep(1)") do |rs|
      rs.each do
        foo =  rs.read(String)
        timer =  rs.read(Bytes)
        pp foo, timer
      end
      puts "called 1"
    end
    puts "Not called"
  end
  spawn do
    db.query("select 'bar'") do |rs|
      rs.each do
        bar =  rs.read(String)
        pp bar
      end
      puts "called 2"
    end
    puts "called 3"
  end
  sleep 2
end
```

Output:

```bash
"PG bar" # => "PG bar"
"PG foo" # => "PG foo"
bar # => "bar"
called 2
called 3
Unhandled exception in spawn:
Closed stream (IO::Error)
0x10943c945: *CallStack::unwind:Array(Pointer(Void)) at ??
0x10943c8e1: *CallStack#initialize:Array(Pointer(Void)) at ??
0x10943c8b8: *CallStack::new:CallStack at ??
0x10942dfd1: *raise<IO::Error>:NoReturn at ??
0x1094e9196: *TCPSocket+@IO#check_open:Nil at ??
0x1094e902b: *TCPSocket+@IO::Buffered#read<Slice(UInt8)>:Int32 at ??
0x1094e8f75: *TCPSocket+@IO#read_fully?<Slice(UInt8)>:(Int32 | Nil) at ??
0x1094e8ec3: *TCPSocket+@IO#read_fully<Slice(UInt8)>:Int32 at ??
0x1094c1584: *MySql::ReadPacket#initialize<TCPSocket+, MySql::Connection>:UInt8 at ??
0x1094c150b: *MySql::ReadPacket::new<TCPSocket+, MySql::Connection>:MySql::ReadPacket at ??
0x1094e187c: *MySql::Connection#build_read_packet:MySql::ReadPacket at ??
0x1095212f0: *MySql::ResultSet#move_next:Bool at ??
0x109521a8e: *MySql::ResultSet#do_close:Nil at ??
0x109524922: *DB::ResultSet+@DB::Disposable#close:(Bool | Nil) at ??
0x10943acb0: ~procProc(Nil)@worksheets/test.cr:29 at ??
0x1094559e4: *Fiber#run:(IO::FileDescriptor | Nil) at ??
0x109433899: ~proc2Proc(Fiber, (IO::FileDescriptor | Nil))@/usr/local/Cellar/crystal-lang/0.22.0/src/fiber.cr:29 at ??
```

**Comment by benoist** [comment]
I've been able to debug this further.

MySql::ResultSet#do_close
```crystal
def do_close
    super

    while move_next
    end

    if row_packet = @row_packet
      row_packet.discard
    end
end
```

With the above code when the `@row_packet` is still filled, the exception is thrown, because the call to `super` closes the connection via `statement.release_connection`

This causes the connection to be closed when the `max_idl_pool_size` is reached. The default is 1 so in the above script the connection will be closed.

When this occurs the connection is closed before the row_packet is discarded, which means the remaining bytes cannot be skipped in the closed socket, hence the `Closed stream (IO::Error)`

I was able to get rid of the exception by calling super in an ensure block, this causes the connection to be closed after discarding the remaining row_packet.

```crystal
def do_close
    while move_next
    end

    if row_packet = @row_packet
      row_packet.discard
    end
ensure
    super
end
```
